### PR TITLE
feat: handle native api changes

### PR DESCRIPTION
**Pull request by dev-marek**
## Issue

https://gravitee.atlassian.net/browse/APIM-7748

## Description

Added handling of NativePlans

## TO DO

- [x] Deuglify
- [x] Fix tests
- [x] Test if happy path (both http and native) works

## Additional context

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-wuyefukqac.chromatic.com)

### CODE REVIEW

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiPlansResourceTest.java
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiPlansResourceTest.java:1]: Maybe add a test here for native plan updates?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/v4/impl/PlanServiceImpl.java
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/v4/impl/PlanServiceImpl.java:1]: Why did you change the old service if it is not used by the use case?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/plan/domain_service/UpdatePlanDomainService.java
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/plan/domain_service/UpdatePlanDomainService.java:1]: To help readability, what do you think of instead of lots of if/then, just have two private methods to handle native vs non-native?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/plan/use_case/UpdatePlanUseCase.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        var planDefinitionV4 = newPlan.getPlanDefinitionV4();
+        planDefinitionV4.setSelectionRule(updatePlan.getSelectionRule());
+        planDefinitionV4.getSecurity().setConfiguration(updatePlan.getSecurity().getConfiguration());
+
+        return newPlan;
+    }
+
+    @Builder
+    public record Input(
+        UpdatePlanEntity planToUpdate,
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/plan/use_case/UpdatePlanUseCase.java]: It seems strange to me to use a `rest.api.model` object here. It might be cleaner to create a onion-specific class that can handle the update of a plan. 
This will then allow for independence from the old services and can minimize changes in this PR.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/plan/use_case/UpdatePlanUseCaseTest.java
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/plan/use_case/UpdatePlanUseCaseTest.java:1]: Maybe add a test for native plans here?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-model/src/main/java/io/gravitee/rest/api/model/v4/plan/UpdatePlanEntity.java [hunk 2]
```diff
@@ -80,8 +79,6 @@ public class UpdatePlanEntity {
 
     private String selectionRule;
 
-    private List<Flow> flows;
-
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-model/src/main/java/io/gravitee/rest/api/model/v4/plan/UpdatePlanEntity.java]: I'm worried that changing this file could have repercussions in the old, non-onion part of the application. 
Perhaps we can use another class/create another class for all of the update information for a plan?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/plan/use_case/UpdatePlanUseCase.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        if (!planEntity.getApiId().equals(input.apiId)) {
+            throw new PlanNotFoundException(input.planToUpdate.getId());
+        }
+
+        var updatedEntity = input.planToUpdate.applyTo(planEntity);
+
+        var api = apiCrudService.get(input.apiId);
+        List<? extends AbstractFlow> flows = input.flowProvider.apply(api);
+
+        var updated = updatePlanDomainService.update(updatedEntity, flows, Map.of(), api, input.auditInfo);
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/plan/use_case/UpdatePlanUseCase.java:57]: Should the `existingPlanStatuses` be an empty Map by default?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/plan/use_case/UpdatePlanUseCaseTest.java
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/plan/use_case/UpdatePlanUseCaseTest.java:1]: Maybe add a test for native plan here?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/plan/use_case/UpdatePlanUseCase.java (hunk 3 above)
Comment (dev-marek) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/plan/use_case/UpdatePlanUseCase.java:57]: That was the original behavior from PlanService

Comment (jourdiw) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/plan/use_case/UpdatePlanUseCase.java:57]: Seeing as the `status` in the plan definitions is not updated when calling the `apply(Plan oldPlan)` in the `PlanUpdates`, the status check is not necessarily valid in `UpdatePlanDomainService`. 
All that to say, it's fine to pass `Map.of()` here for the existing plan statuses.

**Review by jourdiw — APPROVED** [review]
